### PR TITLE
Refactoring of MOAIFreeTypeFont

### PR DESCRIPTION
**Pull request by IkeBart**
I added a couple of methods to reduce repeated lines of code in the source code for MOAIFreeTypeFont.  I also did some reorganization of the list of method declarations in the header.
## TODO:
- [x] Check leaks.

Review:
- [ ] @NachoSoto
- [x] someone else

### CODE REVIEW

**Comment by NachoSoto** [comment]
Could you run this with Instruments and see if there are any leaks? :)

**Comment by aab29** [comment]
screen shot 2013-12-16 at 11 34 48 am
Not bad!

Like @NachoSoto said, could you run this through Instruments a couple times?  Please also look for ways to optimize for speed this week, and profile the optimizations with Instruments.  The biggest problem we have with text right now is just that it's too slow in general.  :speedboat: 

While we're at it...
We're getting occasional crashes where either the texture or the optimal size value returned by `MOAIFreeTypeFont`/`MOAITextRenderer` will be `nil`.  It's probably just something we're doing wrong... like passing `nil` content or empty strings to Moai.  @IkeBart are there any sensible checks that you can think of that could be added to `MOAIFreeTypeFont` or `MOAITextRenderer` to prevent it from failing silently or just returning `nil`?  Be careful, because I've heard that FreeType does silly things.  For example, its version of "`assert`" is actually a call to `exit()`, which just shuts down the app with no stack trace.  If it fails, we'd like it to fail with as much context as possible.  If this is difficult or wouldn't make sense, are there any checks we could do in Lua instead?  Thanks.  :train:

**Comment by NachoSoto** [comment]
Yeah, let's change `CHECK_ERROR` to call `assert` instead of `abort`, **BUT** make sure that we're not compiling with `NDEBUG`.

**Comment by kstenerud** [comment]
Actually, you could just make it call `abort()` instead of `exit()`. Then it doesn't matter whether NDEBUG is on or off.

**Comment by NachoSoto** [comment]
Would the JVM capture `abort` and throw a Java exception?

**Comment by kstenerud** [comment]
I doubt it. The JVM would have to capture SIGABRT, which could be thrown by anything in any C library. That would get messy real quick, especially since you can't run Java code in an async-safe manner.

**Comment by NachoSoto** [comment]
Then let's not use `abort` either, let's throw an exception.

**Comment by kstenerud** [comment]
Oh yeah right. This is C++ code :P

**Comment by aab29** [comment]
@IkeBart Have you run this through Instruments yet?  :trumpet: :violin: :guitar:

I think we should keep our pull requests small, so let's merge this as soon as possible, and we can focus on adding more assertions and optimizing in future pull requests.

**Comment by IkeBart** [comment]
@aab29 I have not run it through instruments yet because I get an Apple Mach-O Linker Error when I use the Profile option from Xcode's Product menu.

The error reads:

```
ld: file not found: /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-tapjoy.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)

File not found: /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-tapjoy.a
```

Can someone else try running the profiling on this branch using Instruments?

**Comment by NachoSoto** [comment]
`libmoai-ios-tapjoy` should be gone :/ Can you share the rest of the output?

**Comment by IkeBart** [comment]
@NachoSoto Here is a more complete transcript of the issues I got when trying to run with the Profile option.  The first three issues are warnings, the last is an error.

```

Check dependencies
Missing dependency target "libmoai-ios-facebook (from libmoai.xcodeproj)"
Missing dependency target "libmoai-ios-tapjoy (from libmoai.xcodeproj)"
Missing dependency target "libmoai-iphone-untz (from libmoai.xcodeproj)"


Ld /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/moai.app/moai normal i386
    cd /Users/isaacbarrett/GitProjects/Moai/mindsnacks-moai-dev/xcode/ios
    setenv IPHONEOS_DEPLOYMENT_TARGET 4.0
    setenv PATH "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin"
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -arch i386 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator6.1.sdk -L/Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator -L/Users/isaacbarrett/GitProjects/Moai/mindsnacks-moai-dev/xcode/ios/../../3rdparty/crittercismiOS-3.3.3/CrittercismSDK-crashonly -L/Users/isaacbarrett/GitProjects/Moai/mindsnacks-moai-dev/xcode/ios/../../3rdparty/facebookiOS-3.0.6.b -F/Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator -filelist /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Intermediates/MoaiSample.build/Release-iphonesimulator/moai.build/Objects-normal/i386/moai.LinkFileList -Xlinker -objc_abi_version -Xlinker 2 -force_load /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-tapjoy.a -fobjc-link-runtime -Xlinker -no_implicit_dylibs -mios-simulator-version-min=4.0 -lfacebook_ios_sdk -lCrittercismCrashOnly_v3_3_3 /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios.a /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-3rdparty.a -lmoai-ios-facebook /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-luaext.a -lmoai-ios-tapjoy -lmoai-ios-untz /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-zlcore.a -framework AudioToolbox -framework CoreAudio -framework CoreGraphics -framework CoreLocation -framework CoreTelephony -framework Foundation -weak_framework GameKit -framework MediaPlayer -framework MessageUI -framework OpenGLES -framework QuartzCore -framework StoreKit -framework SystemConfiguration -framework Twitter -framework UIKit -o /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/moai.app/moai

ld: file not found: /Users/isaacbarrett/Library/Developer/Xcode/DerivedData/MoaiSample-gsmqfxqupbmnziezvidnrzyzfxcn/Build/Products/Release-iphonesimulator/libmoai-ios-tapjoy.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by NachoSoto** [comment]
All those What scheme are you building (profiling) when you get this error?

**Comment by NachoSoto** [comment]
I just realized the schemes are still there. I'm gonna push a commit to mindsnacks-dev.

**Comment by NachoSoto** [comment]
Okay, I just pushed that.
In any case, you were building the wrong scheme, make sure you build libmoai-osx.

**Comment by aab29** [comment]
Can he build libmoai-osx ?  He doesn't have access to Achilles.

**Comment by NachoSoto** [comment]
He can create a sample app that links against `libmoai-osx` and profile it there. How is he testing this code?

**Comment by IkeBart** [comment]
I am testing it from Xcode by hitting Command-R on the keyboard or using the Run option from the Product menu.  I tried using the Profile option from the Product menu on the main branch `mindsnacks-dev` and got the same error.  I was building the scheme called 'moai' with profile.

**Comment by NachoSoto** [comment]
What project are you building though?

**Comment by IkeBart** [comment]
> What project are you building though?

I'm building the project with the filename MoaiSample.xcodeproj in the xcode/ios directory.  `<moai-dev>/xcode/ios/MoaiSample.xcodeproj`

**Comment by NachoSoto** [comment]
We had never fixed that project. I just pushed a commit to `mindsnacks-dev` that should allow you to compile that project again :)

**Comment by aab29** [comment]
@IkeBart Were you able to figure it out?

**Comment by IkeBart** [comment]
Yes, I was able to get Instruments to run by selecting the Profile option after pulling in the changes made to the main branch.

**Comment by aab29** [comment]
Were there any leaks?

**Comment by IkeBart** [comment]
In the quick test, I did not detect any leaks.  But I can't tell for sure without testing it again.  I'm now having trouble rebuilding it after editing the source file.  After running my MOAI test programs once with Instruments, it still runs the same program without checking the source, even after I make changes to the Lua source.  Also, the debug messages don't appear in the console log like they usually do.  It looks like it's in release mode, not debug mode.  Even after rebuilding after cleaning, it still runs the old Lua code.

**Comment by IkeBart** [comment]
After restarting my computer, deleting the Moai app on the simulator, cleaning the project and rebuilding, all I'm getting now is a blank screen whenever I try to run it.  In theory, it should run main.lua and the specified source files with `require()`, but the blank screen with a blank console log is all I get.  I did some stepping through the code and it appears that "main.lua" does not exist according to the check in `AKURunScript()` as defined in AKU.cpp.  AKURunScript() was called by`[MoaiView run:]` which in turn was called by `[MoaiAppDelegate application:didFinishLaunchingWithOptions:]`.

If only I could figure out a way to make the Lua source code exist again over in the simulator app.

**Comment by NachoSoto** [comment]
Is the source a folder reference in Xcode? If so you're going to have to remove the product, clean, and compile again after changing the LUA source.
You can profile in Debug mode by editing the scheme:

screen shot 2013-12-19 at 10 08 20 am

**Comment by NachoSoto** [comment]
I'm not familiar with this project. Where is the `main.lua` file supposed to be?

**Comment by IkeBart** [comment]
> I'm not familiar with this project. Where is the `main.lua` file supposed to be?

The `main.lua` file is supposed to be in the `xcode/ios/lua` directory, which is set up as a folder reference in `xcode/ios/MoaiSample.xcodeproj`.

**Comment by NachoSoto** [comment]
Looks like I mistakenly removed it [here](https://github.com/mindsnacks/moai-dev/commit/b9f4ad280e905d32d2b1ebd110662372ead53511#diff-86cfac4e28ee75f3e8658cc6d37ccc86L29). Could you add it again? :)

**Comment by aab29** [comment]
How's this going, @IkeBart ?

**Comment by aab29** [comment]
@IkeBart bump.  :red_car:

**Comment by IkeBart** [comment]
I call my latest update a work in progress because I'm still getting errors when trying to build the `moai` scheme of `xcode/osx/MoaiSample.xcodeproj`.  The errors are for undefined symbols for architecture i386.  The symbols are functions declared in `src/aku/AKU.h` and defined in `src/aku/AKU.cpp`.  The undefined symbols are referenced in GlutHost.o, probably related to `src/hosts/GlutHost.cpp`.

**Comment by IkeBart** [comment]
The OSX Xcode project builds without errors now.  The next question is how do I test OSX moai with Instruments?  Also, what should I be measuring with Instruments?

**Comment by kstenerud** [comment]
When I was doing some exploratory coding in the font system, I would make a simple moai game that made a bunch of labels (loop 100 times, making a label with some text and the iteration number appended), then check instruments to see where it was spending most of its time. Do that with the old code and the new code and you'll see more clearly how you've affected performance.

File: src/hosts/GlutHost.cpp [hunk 1]
```diff
@@ -399,8 +399,8 @@ void GlutRefreshContext () {
 	#endif
 	
 	#ifdef GLUTHOST_USE_LUAEXT
-		AKUExtLoadLuacrypto ();
-		AKUExtLoadLuacurl ();
+		//AKUExtLoadLuacrypto ();
+		//AKUExtLoadLuacurl ();
```
Comment (NachoSoto) [inline src/hosts/GlutHost.cpp]: Can we remove this instead of commenting out?

File: xcode/libmoai/libmoai.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -8143,12 +8155,14 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 5751E6D018565F8B00775234 /* libmoai-osx-debug.xcconfig */;
 			buildSettings = {
+				ARCHS = "$(ARCHS_STANDARD_32_BIT)";
```
Comment (NachoSoto) [inline xcode/libmoai/libmoai.xcodeproj/project.pbxproj]: The `ARCHS` are defined in the `xcconfig` files.

**Comment by aab29** [comment]
@IkeBart Can you address the remaining comments please?

**Comment by aab29** [comment]
Is there anything that still needs to be done here?

**Comment by IkeBart** [comment]
From my testing in Instruments, I found that calls to `FT_Load_Glyph` take up most of the CPU time in `MOAIFreeTypeFont::RenderTexture()`, at least on the IOS simulator.  I have not seen any significant difference in the total amount of time spent in `MOAIFreeTypeFont::RenderTexture()` between this branch and the `mindsnacks-dev` branch.  I have yet to compare `MOAIFreeTypeFont::RenderTextureSingleLine()` and `MOAIFreeTypeFont::OptimalSize()` between the branches.

**Comment by aab29** [comment]
OK, if you're sure nothing is leaking, we should merge this.  :shipit:

In a separate branch, could you please work on optimizing this some more?  I'm betting you can find a smart way to minimize the number of calls to `FT_Load_Glyph`.

**Comment by aab29** [comment]
@NachoSoto Did you want to look at this?  If there's nothing wrong, I'd like to get it merged finally.  :rabbit:

**Comment by NachoSoto** [comment]
:+1:

**Pull request merged**